aiohttp/helpers.py

```diff
@@ -797,9 +797,38 @@ def set_result(fut: "asyncio.Future[_T]", result: _T) -> None:
         fut.set_result(result)
 
 
-def set_exception(fut: "asyncio.Future[_T]", exc: BaseException) -> None:
-    if not fut.done():
-        fut.set_exception(exc)
+class ErrorableMixin:
+    _EXC_SENTINEL = BaseException()
+
+    def set_exception(
+        self,
+        exc: BaseException,
+        exc_cause: BaseException = _EXC_SENTINEL,
+    ):
+        raise NotImplementedError
+
+
+def set_exception(
+    fut: "asyncio.Future[_T] | ErrorableMixin",
```

Comment:
@Dreamsorcerer so with https://github.com/aio-libs/aiohttp/pull/8089#discussion_r1469722650, do you mean that doing as follows would mean that I'd simply drop the inheritance from other modules even from `StreamReader`?

```suggestion
class ErrorableProtocol(asyncio.Protocol):
    _EXC_SENTINEL = BaseException()

    def set_exception(
        self,
        exc: BaseException,
        exc_cause: BaseException = _EXC_SENTINEL,
    ):
        raise NotImplementedError


def set_exception(
    fut: "asyncio.Future[_T] | ErrorableProtocol",
```

Also, did you mean `asyncio.Protocol` or `aiohttp.base_protocol.BaseProtocol`? Or maybe, `asyncio.BaseProtocol`?

Reply:
Uhh, `typing.Protocol`. :P

Reply:
You basically define a stub with Protocol, and then the type checker checks for compatibility with the interface of the stub (i.e. duck typing).

As it's a stub, I'd include the `...` as in my comment earlier.
`_EXC_SENTINEL` will need to be moved to a separate variable or something.

Reply:
Ah, so just

```suggestion
class ErrorableProtocol(Protocol):
    _EXC_SENTINEL = BaseException()

    def set_exception(
        self,
        exc: BaseException,
        exc_cause: BaseException = _EXC_SENTINEL,
    ):
        raise NotImplementedError


def set_exception(
    fut: "asyncio.Future[_T] | ErrorableProtocol",
```

Is the existence of `_EXC_SENTINEL` on the class going to be a problem?

Reply:
> Is the existence of `_EXC_SENTINEL` on the class going to be a problem?

Yep, it'll be part of the stub it's looking for. I don't think there's any need to have it as part of a class.

Reply:
I guess, I'll need to move it to the module level, then.

Reply:
Yeah, most likely the class won't need to referenced anywhere else, so you'll still only be importing one thing.

Reply:
Yep. Seems to work.